src/platform/Darwin/DnssdInterfaceInfos.cpp

```diff
@@ -0,0 +1,183 @@
+{
+    return !(flags & IFF_POINTOPOINT) && (flags & IFF_RUNNING) && (flags & IFF_MULTICAST);
+}
+
+bool InterfaceInfos::HasValidType(Inet::IPAddressType addressType, const sa_family_t addressFamily) const
+{
+    bool useIPv4 = addressType == chip::Inet::IPAddressType::kIPv4 || addressType == chip::Inet::IPAddressType::kAny;
+    bool useIPv6 = addressType == chip::Inet::IPAddressType::kIPv6 || addressType == chip::Inet::IPAddressType::kAny;
+
+    return (useIPv6 && addressFamily == AF_INET6) || (useIPv4 && addressFamily == AF_INET);
+}
+
+bool InterfaceInfos::IsValidInterfaceId(uint32_t targetInterfaceId, unsigned int currentInterfaceId) const
+{
+    return targetInterfaceId == kDNSServiceInterfaceIndexAny || targetInterfaceId == currentInterfaceId;
+}
+
+void InterfaceInfos::LogDetail() const
+{
+    char interfaceName[IFNAMSIZ] = {};
+    VerifyOrReturn(if_indextoname(mInterfaceId, interfaceName) != nullptr);
+
+    if (IsIPv4())
+    {
+        char addr[INET_ADDRSTRLEN] = {};
+        inet_ntop(AF_INET, GetIPv4(), addr, sizeof(addr));
+        ChipLogDetail(Discovery, "\tInterface: %s (%u) ipv4: %s", interfaceName, mInterfaceId, addr);
+    }
+    else
+    {
+        char addr[INET6_ADDRSTRLEN] = {};
+        inet_ntop(AF_INET6, GetIPv6(), addr, sizeof(addr));
+        ChipLogDetail(Discovery, "\tInterface: %s (%u) ipv6: %s", interfaceName, mInterfaceId, addr);
+    }
+}
+
+bool InterfacesInfos::Init(const char * hostname, chip::Inet::IPAddressType addressType, uint32_t interfaceId)
+{
+    mHostName = hostname;
+
+    // When the target interface is kDNSServiceInterfaceIndexLocalOnly, there is no need to map the hostname to
+    // an IP address.
+    VerifyOrReturnValue(interfaceId != kDNSServiceInterfaceIndexLocalOnly, true);
+
+    struct ifaddrs * ifap;
+    VerifyOrReturnValue(getifaddrs(&ifap) >= 0, false);
+
+    struct ifaddrs * ifa_start = ifap;
+    for (struct ifaddrs * ifa = ifa_start; ifa; ifa = ifa->ifa_next)
+    {
+        InterfaceInfos interfaceInfos;
+        if (interfaceInfos.Init(ifa, addressType, interfaceId))
+        {
+            mInterfaceInfos.push_back(interfaceInfos);
+        }
+    }
+
+    freeifaddrs(ifap);
+
+    return mInterfaceInfos.size();
+}
+
+CHIP_ERROR InterfacesInfos::RegisterRecords(RegisterContext * registerCtx)
+{
+    DNSServiceRef sdRef;
+    auto err = DNSServiceCreateConnection(&sdRef);
+    VerifyOrReturnError((kDNSServiceErr_NoError == err || sdRef != nullptr), CHIP_ERROR_INTERNAL);
+
+    RegisterRecordContext * sdCtx = chip::Platform::New<RegisterRecordContext>(registerCtx);
+    VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
+
+    ChipLogDetail(Discovery, "Mdns: Mapping %s to:", mHostName.c_str());
+
+    auto count = mInterfaceInfos.size();
+    for (size_t i = 0; i < count; i++)
+    {
+        err = mInterfaceInfos[i].Register(sdRef, sdCtx, mHostName.c_str());
+        VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
```

Comment:
Again, I'm not sure we should do that `Finalize(err)`, unless _all_ the interfaces fail.